File: python system/inventory_system/services/transaction_service.py
import uuid
from typing import Optional, Tuple
from datetime import datetime
from models.inventory_item import InventoryItem
from models.transaction import Transaction, TransactionType
from repositories.inventory_repository import InventoryRepository
from repositories.transaction_repository import TransactionRepository
from services.audit_logger import AuditLogger


class TransactionException(Exception):
    def __init__(self, message: str):
        self.message = message
        super().__init__(message)


class TransactionResult:
    def __init__(
        self,
        success: bool,
        error: Optional[str] = None,
        transaction: Optional[Transaction] = None,
        updated_item: Optional[InventoryItem] = None,
    ):
        self.success = success
        self.error = error
        self.transaction = transaction
        self.updated_item = updated_item

    @classmethod
    def ok(cls, transaction: Transaction, updated_item: InventoryItem) -> "TransactionResult":
        return cls(success=True, transaction=transaction, updated_item=updated_item)

    @classmethod
    def fail(cls, error: str) -> "TransactionResult":
        return cls(success=False, error=error)


class TransactionService:
    def __init__(
        self,
        inventory_repo: InventoryRepository,
        transaction_repo: TransactionRepository,
        audit_logger: AuditLogger,
    ):
        self._inventory_repo = inventory_repo
        self._transaction_repo = transaction_repo
        self._audit_logger = audit_logger

    def stock_in(
        self,
        item_id: str,
        quantity: int,
        reference_id: Optional[str] = None,
        performed_by: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> TransactionResult:
        if quantity <= 0:
            return TransactionResult.fail("Quantity must be greater than zero.")

        item = self._inventory_repo.find_by_id(item_id)
        if item is None:
            return TransactionResult.fail(f"Item not found: {item_id}")

        updated_item = item.copy_with(
            quantity=item.quantity + quantity,
            last_updated=datetime.now(),
        )

        tx = Transaction(
            id=str(uuid.uuid4()),
            timestamp=datetime.now(),
            type=TransactionType.STOCK_IN,
            item_id=item.id,
            item_sku=item.sku,
            quantity_delta=quantity,
            quantity_before=item.quantity,
            quantity_after=updated_item.quantity,
            unit_cost=item.unit_cost,
            reference_id=reference_id,
            performed_by=performed_by,
            notes=notes,
        )

        self._inventory_repo.save(updated_item)
        self._transaction_repo.save(tx)
        self._audit_logger.log_stock_in(item, updated_item, user_id=performed_by, notes=notes)

        return TransactionResult.ok(tx, updated_item)

    def stock_out(
        self,
        item_id: str,
        quantity: int,
        reference_id: Optional[str] = None,
        performed_by: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> TransactionResult:
        if quantity <= 0:
            return TransactionResult.fail("Quantity must be greater than zero.")

        item = self._inventory_repo.find_by_id(item_id)
        if item is None:
            return TransactionResult.fail(f"Item not found: {item_id}")

        if item.quantity < quantity:
            return TransactionResult.fail(
                f"Insufficient stock. Available: {item.quantity}, Requested: {quantity}"
            )

        updated_item = item.copy_with(
            quantity=item.quantity - quantity,
            last_updated=datetime.now(),
        )

        tx = Transaction(
            id=str(uuid.uuid4()),
            timestamp=datetime.now(),
            type=TransactionType.STOCK_OUT,
            item_id=item.id,
            item_sku=item.sku,
            quantity_delta=-quantity,
            quantity_before=item.quantity,
            quantity_after=updated_item.quantity,
            unit_cost=item.unit_cost,
            reference_id=reference_id,
            performed_by=performed_by,
            notes=notes,
        )

        self._inventory_repo.save(updated_item)
        self._transaction_repo.save(tx)
        self._audit_logger.log_stock_out(item, updated_item, user_id=performed_by, notes=notes)

        return TransactionResult.ok(tx, updated_item)

    def adjust(
        self,
        item_id: str,
        new_quantity: int,
        performed_by: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> TransactionResult:
        if new_quantity < 0:
            return TransactionResult.fail("Quantity cannot be negative.")

        item = self._inventory_repo.find_by_id(item_id)
        if item is None:
            return TransactionResult.fail(f"Item not found: {item_id}")

        delta = new_quantity - item.quantity
        updated_item = item.copy_with(
            quantity=new_quantity,
            last_updated=datetime.now(),
        )

        tx = Transaction(
            id=str(uuid.uuid4()),
            timestamp=datetime.now(),
            type=TransactionType.ADJUSTMENT,
            item_id=item.id,
            item_sku=item.sku,
            quantity_delta=delta,
            quantity_before=item.quantity,
            quantity_after=new_quantity,
            unit_cost=item.unit_cost,
            performed_by=performed_by,
            notes=notes or "Manual adjustment",
        )

        self._inventory_repo.save(updated_item)
        self._transaction_repo.save(tx)
        self._audit_logger.log_adjustment(item, updated_item, user_id=performed_by, notes=notes)

        return TransactionResult.ok(tx, updated_item)

File: dart/transaction_system/dart_application_1/bin/transaction_service.dart
// lib/services/transaction_service.dart

import 'package:uuid/uuid.dart';
import 'models/inventory_item.dart';
import 'models/transaction.dart';
import 'inventory_repository.dart';
import 'transaction_repository.dart';
import 'audit_logger.dart';

class TransactionException implements Exception {
  final String message;
  TransactionException(this.message);

  @override
  String toString() => 'TransactionException: $message';
}

class TransactionResult {
  final bool success;
  final String? error;
  final Transaction? transaction;
  final InventoryItem? updatedItem;

  const TransactionResult.ok({
    required this.transaction,
    required this.updatedItem,
  })  : success = true,
        error = null;

  const TransactionResult.fail(this.error)
      : success = false,
        transaction = null,
        updatedItem = null;
}

class TransactionService {
  final InventoryRepository _inventoryRepo;
  final TransactionRepository _transactionRepo;
  final AuditLogger _auditLogger;
  final Uuid _uuid;

  TransactionService({
    required InventoryRepository inventoryRepo,
    required TransactionRepository transactionRepo,
    required AuditLogger auditLogger,
  })  : _inventoryRepo = inventoryRepo,
        _transactionRepo = transactionRepo,
        _auditLogger = auditLogger,
        _uuid = const Uuid();

  // ── Stock In ──────────────────────────────────────────────────────
  TransactionResult stockIn({
    required String itemId,
    required int quantity,
    String? referenceId,
    String? performedBy,
    String? notes,
  }) {
    if (quantity <= 0) {
      return const TransactionResult.fail('Quantity must be greater than zero.');
    }

    final item = _inventoryRepo.findById(itemId);
    if (item == null) {
      return TransactionResult.fail('Item not found: $itemId');
    }

    final updatedItem = item.copyWith(
      quantity: item.quantity + quantity,
      lastUpdated: DateTime.now(),
    );

      final tx = Transaction(
        id: _uuid.v4(),
        timestamp: DateTime.now(),
        type: TransactionType.stockIn,
      itemId: item.id,
      itemSku: item.sku,
      quantityDelta: quantity,
      quantityBefore: item.quantity,
      quantityAfter: updatedItem.quantity,
      unitCost: item.unitCost,
      referenceId: referenceId,
      performedBy: performedBy,
      notes: notes,
    );

    // Atomic: save both item + transaction + audit log
    _inventoryRepo.save(updatedItem);
    _transactionRepo.save(tx);
    _auditLogger.logStockIn(item, updatedItem, userId: performedBy, notes: notes);

    return TransactionResult.ok(transaction: tx, updatedItem: updatedItem);
  }

  // ── Stock Out ─────────────────────────────────────────────────────
  TransactionResult stockOut({
    required String itemId,
    required int quantity,
    String? referenceId,
    String? performedBy,
    String? notes,
  }) {
    if (quantity <= 0) {
      return const TransactionResult.fail('Quantity must be greater than zero.');
    }

    final item = _inventoryRepo.findById(itemId);
    if (item == null) {
      return TransactionResult.fail('Item not found: $itemId');
    }

    if (item.quantity < quantity) {
      return TransactionResult.fail(
        'Insufficient stock. Available: ${item.quantity}, Requested: $quantity',
      );
    }

    final updatedItem = item.copyWith(
      quantity: item.quantity - quantity,
      lastUpdated: DateTime.now(),
    );

    final tx = Transaction(
      id: _uuid.v4(),
      timestamp: DateTime.now(),
      type: TransactionType.stockOut,
      itemId: item.id,
      itemSku: item.sku,
      quantityDelta: -quantity,
      quantityBefore: item.quantity,
      quantityAfter: updatedItem.quantity,
      unitCost: item.unitCost,
      referenceId: referenceId,
      performedBy: performedBy,
      notes: notes,
    );

    _inventoryRepo.save(updatedItem);
    _transactionRepo.save(tx);
    _auditLogger.logStockOut(item, updatedItem, userId: performedBy, notes: notes);

    return TransactionResult.ok(transaction: tx, updatedItem: updatedItem);
  }

  // ── Manual Adjustment ─────────────────────────────────────────────
  TransactionResult adjust({
    required String itemId,
    required int newQuantity,
    String? performedBy,
    String? notes,
  }) {
    if (newQuantity < 0) {
      return const TransactionResult.fail('Quantity cannot be negative.');
    }

    final item = _inventoryRepo.findById(itemId);
    if (item == null) {
      return TransactionResult.fail('Item not found: $itemId');
    }

    final delta = newQuantity - item.quantity;
    final updatedItem = item.copyWith(
      quantity: newQuantity,
      lastUpdated: DateTime.now(),
    );

    final tx = Transaction(
      id: _uuid.v4(),
      timestamp: DateTime.now(),
      type: TransactionType.adjustment,
      itemId: item.id,
      itemSku: item.sku,
      quantityDelta: delta,
      quantityBefore: item.quantity,
      quantityAfter: newQuantity,
      unitCost: item.unitCost,
      performedBy: performedBy,
      notes: notes ?? 'Manual adjustment',
    );

    _inventoryRepo.save(updatedItem);
    _transactionRepo.save(tx);
    _auditLogger.logAdjustment(item, updatedItem, userId: performedBy, notes: notes);

    return TransactionResult.ok(transaction: tx, updatedItem: updatedItem);
  }
}
